feat: generalize to arbitrary bitvec width `zeroextend_bigger_smaller` and `truncate_of_concat_is_lsb`

luisacicolini opened the pull request:
### Description:

This PR extends `truncate_of_concat_is_lsb_64` to arbitrary-width bitvectors. 

### Testing:

What tests have been run? Did `make all` succeed for your changes? Yes
Was conformance testing successful on an Aarch64 machine? yes 

### License:

By submitting this pull request, I confirm that my contribution is
made under the terms of the Apache 2.0 license.

shigoel commented on Proofs/SHA512/SHA512_block_armv8_rules.lean:
```diff
@@ -134,17 +134,25 @@ theorem concat_of_rsh_is_msb_128 (x y : BitVec 64) :
   BitVec.zeroExtend 64 x := by
   bv_omega
 
+theorem zeroextend_bigger_smaller (w : Nat) (x : BitVec w) :
```
This could be generalized further. In fact, the lemma we really need already exists in Lean --  `BitVec.zeroExtend_zeroExtend_of_le`. We should really delete `zeroextend_bigger_smaller_64` and `zeroextend_bigger_smaller`, and use that lemma instead.

shigoel commented on Proofs/SHA512/SHA512_block_armv8_rules.lean:
```diff
@@ -134,17 +134,25 @@ theorem concat_of_rsh_is_msb_128 (x y : BitVec 64) :
--- (FIXME) Generalize to arbitrary-length bitvecs.
+theorem truncate_of_concat_is_lsb (w : Nat) (x y : BitVec w) :
+  BitVec.zeroExtend w (x ++ y) = y := by
+  ext
+  simp [BitVec.leftshift_n_or_mod_2n]
```
Thanks for this! In LNSym, we prefer using `simp only` wherever possible, so could you please fix that? Note that you can also use `BitVec.truncate_append` for a smaller/cleaner proof here.

Also, I acknowledge that this file has badly named lemmas (it precedes Lean's BitVec lemma dev) --- this lemma should've really used `zeroExtend` and `append` in the name.

shigoel reviewed (commented):
Thanks for the PR, @luisacicolini!

A high-level comment: could you please use [LNSym's PR template](https://github.com/leanprover/LNSym/blob/main/.github/pull_request_template.md) for this PR's description?

luisacicolini commented:
thank you for reviewing this @shigoel! Everything should be fixed now. I removed `zeroextend_bigger_smaller` and added a comment pointing to the correct arbitrary-width version.

alexkeizer commented on Proofs/SHA512/SHA512_block_armv8_rules.lean:
```diff
@@ -134,12 +134,16 @@ theorem concat_of_rsh_is_msb_128 (x y : BitVec 64) :
   bv_check "lrat_files/Sha512_block_armv8_rules.lean-truncate_of_concat_is_lsb_64-106-2.lrat"
 
--- (FIXME) Generalize to arbitrary-length bitvecs.
+theorem zeroExtend_append_eq (w : Nat) (x y : BitVec w) :
+  BitVec.zeroExtend w (x ++ y) = y := by
+  ext
+  simp only [truncate_append, Nat.le_refl, ↓reduceDIte, zeroExtend_eq]
```
The style in LNSym is currently a bit inconsistent, but standard Lean/Mathlib style is to indent the type/statement of a theorem with 4 spaces (vs 2 spaces for the proof itself).

Then, we can generalize the theorem further: append is a heterogenous operation and can take two bitvectors of different length! Since the value `w` and `v` is determined by the other arguments `x` and `y`, we can make the former implicit arguments.

Finally, you could consider naming this `zeroExtend_append_eq_right`, to draw extra attention to the fact that the value we're "zero-extending" to is exactly the width of the right operand, but it's also perfectly valid to argue that's the logical statement for a theorem about zeroExtend and append, and keep preferring the shorter name, I'll leave that choice up to you.

```suggestion
/-- truncating the result of appending two bitvectors to the width of the right operand gives
exactly the right operand -/
theorem zeroExtend_append_eq {w v : Nat} (x : BitVec w) (y : BitVec v) :
    BitVec.zeroExtend v (x ++ y) = y := by
  ext
  simp only [truncate_append, Nat.le_refl, ↓reduceDIte, zeroExtend_eq]
```

alexkeizer reviewed (dismissed):
Thanks for the contribution! LGTM, but I've left some nits on how the theorem could be generalized further.

shigoel reviewed (approved):
Thanks, @luisacicolini!
